File: Characters/NPC.js
// File: NPC.js
class NPC {
    constructor(name, thear) {
        this.name = name;
        this.behavior = "Idle";
        this.isHungry = false;
        this.isThirsty = false;
        this.isTired = false;
        this.properties = []; // Initialize properties array
        this.thear = thear; // Inject Thear object
    }

    updateBehavior(time, weather) {
        const { Schedule, Behaviors } = this.thear.Bractalia.Nexus.Taverne.NPCs;
        newBehavior = behaviors.waiting; // Default behavior
        // Ensure Thear object and its nested properties exist
        if (typeof time !== 'number' || time < 0 || time > 24) {
            throw new TypeError('Invalid value for time parameter. Time must be a number between 0 and 24.');
        }

        // Validate weather parameter
        if (typeof weather !== 'string') {
            throw new TypeError('Invalid value for weather parameter. Weather must be a string.');
        }
        if (!Thear || !Thear.Bractalia || !Thear.Bractalia.Nexus || !Thear.Bractalia.Nexus.Taverne || !Thear.Bractalia.Nexus.Taverne.NPCs) {
            throw new Error('Missing required dependencies for NPC behavior update.');
        }
        if (time >= Schedule.WakeUp && time < Schedule.WorkStart) {
            newBehavior = Behaviors.Idle;
        } else if (time >= Schedule.WorkStart && time < Schedule.LunchStart) {
            newBehavior = Behaviors.PublicWork;
        } else if (time >= Schedule.LunchStart && time < Schedule.WorkEnd) {
            newBehavior = Behaviors.Socializing;
        } else if (time >= Schedule.WorkEnd && time < Schedule.DinnerStart) {
            newBehavior = Behaviors.Idle;
        } else if (time >= Schedule.DinnerStart && time < Schedule.Bedtime) {
            newBehavior = Behaviors.Socializing;
        } else if (this.isHungry) {
            newBehavior = "Eating";
        } else if (this.isThirsty) {
            newBehavior = "Drinking";
        } else if (this.isTired) {
            newBehavior = "Sleeping";
        } else {
            newBehavior = "Waiting";
        }

        // Weather-based behavior modification
        if (weather === "Stormy") {
            newBehavior = "Seeking Shelter";
        }

        this.behavior = newBehavior; // Update behavior
    }

    interact() {
        console.log(`${this.name} says: Hello there!`);
    }

    ownProperty(property) {
        this.properties.push(property);
    }

    customizeProperty(property, customization) {
        property.addCustomization(customization);
        console.log(`${this.name} has customized ${property.type} at ${property.location}.`);
    }
}

class Property {
    constructor(type, location) {
        this.type = type;
        this.location = location;
        this.customizations = [];
    }

    addCustomization(customization) {
        this.customizations.push(customization);
    }
}

module.exports = { NPC, Property }; // Export NPC and Property classes

File: Characters/NPC.py
# File: NPC.py

class NPC:
    def __init__(self, name, thear):
        self.name = name
        self.behavior = "Idle"
        self.isHungry = False
        self.isThirsty = False
        self.isTired = False
        self.properties = []  # Initialize properties array
        self.thear = thear  # Inject Thear object

    def updateBehavior(self, time, weather):
        Schedule = self.thear.Bractalia.Nexus.Taverne.NPCs.Schedule
        Behaviors = self.thear.Bractalia.Nexus.Taverne.NPCs.Behaviors
        newBehavior = Behaviors.Idle  # Default behavior
        # Ensure Thear object and its nested properties exist
        if not isinstance(time, int) or time < 0 or time > 24:
            raise TypeError('Invalid value for time parameter. Time must be a number between 0 and 24.')

        # Validate weather parameter
        if not isinstance(weather, str):
            raise TypeError('Invalid value for weather parameter. Weather must be a string.')
        if not self.thear or not self.thear.Bractalia or not self.thear.Bractalia.Nexus \
                or not self.thear.Bractalia.Nexus.Taverne or not self.thear.Bractalia.Nexus.Taverne.NPCs:
            raise Error('Missing required dependencies for NPC behavior update.')
        if time >= Schedule.WakeUp and time < Schedule.WorkStart:
            newBehavior = Behaviors.Idle
        elif time >= Schedule.WorkStart and time < Schedule.LunchStart:
            newBehavior = Behaviors.PublicWork
        elif time >= Schedule.LunchStart and time < Schedule.WorkEnd:
            newBehavior = Behaviors.Socializing
        elif time >= Schedule.WorkEnd and time < Schedule.DinnerStart:
            newBehavior = Behaviors.Idle
        elif time >= Schedule.DinnerStart and time < Schedule.Bedtime:
            newBehavior = Behaviors.Socializing
        elif self.isHungry:
            newBehavior = "Eating"
        elif self.isThirsty:
            newBehavior = "Drinking"
        elif self.isTired:
            newBehavior = "Sleeping"
        else:
            newBehavior = "Waiting"

        # Weather-based behavior modification
        if weather == "Stormy":
            newBehavior = "Seeking Shelter"

        self.behavior = newBehavior  # Update behavior

    def interact(self):
        print(f"{self.name} says: Hello there!")

    def ownProperty(self, property):
        self.properties.append(property)

    def customizeProperty(self, property, customization):
        property.addCustomization(customization)
        print(f"{self.name} has customized {property.type} at {property.location}.")

class Property:
    def __init__(self, type, location):
        self.type = type
        self.location = location
        self.customizations = []

    def addCustomization(self, customization):
        self.customizations.append(customization)

# Export NPC and Property classes
__all__ = ['NPC', 'Property']
